### PR TITLE
Add Avatars to users

### LINKED ISSUE #24: Users should have an avatar
This could come from either gravatar or google.

### PR DESCRIPTION
**Pull request by josefwaller**
Automatically search for avatars using the users' email. Can find either gravatars or google profile pictures. If neither is found, defaults back to the current icon. Closes #24.

Example Avatar:
imperilment-avatar

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -57,4 +77,66 @@ def correct_ratio(season = nil)
       0.0
     end
   end
+
+  # Set the avatar type to either default (icon) or Gravatar
```
Comment (cbrunsdon) [inline app/models/user.rb]: Thsi code is better put somewhere else.

It contains much logic (http requests, external services, etc) that are only barely related to what makes a user a user. I'd pull this out into a separate class.

File: app/helpers/avatar_helper.rb [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+      # Get the gravatar
+      image_tag "#{user.avatar_url}&s=30", {width: size, height: size}
+
+    when User.avatar_types[:google]
+      # Get the google image
+      image_tag user.avatar_url, {width: size, height: size}
+
+    else
+      # Logs a warning
+      Rails.logger.warn "%s (id = %d) has no avatar_type! Defaulting to icon." % [user.identifier, user.id]
```
Comment (cbrunsdon) [inline app/helpers/avatar_helper.rb]: This warning isn't providing any value, lots of users will legitimately never have an avatar

Comment (cbrunsdon) [inline app/helpers/avatar_helper.rb]: you can replace this with the explicit check for :default

File: app/models/user.rb [hunk 3]
```diff
@@ -13,6 +13,25 @@ class User < ActiveRecord::Base
...
+
+  # Used to get the Gravatar/Google avatar
+  require "http"
+  require "digest/md5"
+
+  # Used to parse the google image data
+  require "json"
+
+  # Called right before saving
+  def save(*args)
```
Comment (cbrunsdon) [inline app/models/user.rb]: overriding save is pretty rough, I'd suggest either hooking into the before_save callback:

 http://api.rubyonrails.org/classes/ActiveRecord/Callbacks.html

I'd also make sure I wrapped it and caught any exceptions, as it involves HTTP requests and I wouldn't want failed http grabs to stop legit changes from being saved

**Comment by josefwaller** [comment]
Close in favor of #62.

**Pull request closed without merging**